0006-ffmul-revised.md

```diff
@@ -0,0 +1,518 @@
+
+**Expected outcome**: An implementation of the ffmul gate and supporting gadgets which we are confident in to be secure.
+
+## Detailed design
+
+This section is intended to be readable in a self-contained way. However, for a fuller picture, especially of the motivations of many choices made in the design, we recommend to read the [original ffmul RFC](https://o1-labs.github.io/proof-systems/rfcs/foreign_field_mul.html).
+
+Foreign field elements $x$ and other integers of size comparable to $f$ are represented by three native field element limbs $(x_0, x_1, x_2)$, via
+
+$$
+x = x_0 + 2^\ell x_1 + 2^{2\ell} x_2.
+$$
+
+The limb size $\ell = 88$ is chosen to enable efficient $\ell$-bit range checks, so we can prove that $x_i \in [0, 2^\ell)$. This allows us to express any integer $x \in [0, 2^{3\ell}) = [0, 2^{264})$ in limb form, covering important 256-bit finite fields like secp256r1. From now on, we work under the assumption that $f < 2^{3\ell}$.
+
+The ffmul gate represents $a$, $b$, $q$ in 3-limb form. However, $r$ is represented in an alternative compact 2-limb form $(r_{01}, r_2)$ where the two bottom limbs are combined into a single variable,
+
+$$
+r = r_{01} + 2^{2\ell}r_2.
+$$
+
+Our range-check (RC) gates support a "compact mode" where a variable $r_{01}$ is checked to be split like $r_{01} = r_0 + 2^\ell r_{1}$, and $r_0$ and $r_1$ have $\ell$ bits. This will be used to range-check $r$ and expose its individual limbs to other gates. The motivation for using the compact form for $r$ is to free up permutable cells in the gate layout, which is described below.
+
+### Constraints and soundness proof
+
+We will now derive the constraints necessary to prove that
+
+$$\tag{1}
+a b = qf + r
+$$
+
+At the same time, we will prove soundness -- i.e., the constraints we add imply $(1)$.
+
+Let $n$ be the _native_ field modulus (for example, $n$ is one of the Pasta primes). Our first constraint is to check $(1)$ modulo $n$:
+
+$$\tag{C1: native}
+ab = qf + r\mod n
+$$
+
+Note: In the implementation, variables in this constraint are expanded into their limbs, like
+
+$$
+ab = (a_0 + 2^\ell a_1 + 2^{2\ell} a_2)(b_0 + 2^\ell b_1 + 2^{2\ell} b_2).
+$$
+
+Equation $\text{(C1: native)}$ implies that there is an $\varepsilon\in\mathbb{Z}$ such that
+
+$$
+ab - qf - r =  \varepsilon n.
+$$
+
+If the foreign field was small enough, we could prove $|ab - qf - r| < n$ (after adding a few range checks), and we would be finished at this point, because we could conclude that $\varepsilon = 0$. However, for the moduli $f$ we care about, $ab \approx qf \approx f^2$ is much larger than $n$, so we need to do more work.
+
+The broad strategy is to also constrain $(1)$ modulo $2^{3\ell}$, which implies that it holds modulo the product $2^{3\ell}n$ (this the "chinese remainder theorem"). The modulus $2^{3\ell}n$ is large enough that it can replace $n$ in the last paragraph and the argument actually works.
```

Comment:
I fixed the formulation. Since the proof in the end doesn't invoke the CRT (but proves the part of it we need inline, which I think makes it easier to read -- less magic), I don't think we need any more details on the CRT